### PR TITLE
Orca 470 add policy check step runner

### PR DESCRIPTION
**Pull request by msarvar**

### CODE REVIEW

File: server/events/command_runner.go [hunk 1]
```diff
@@ -134,55 +134,89 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(baseRepo models.Repo, headRepo
...
 			// the Atlantis status to be passing for all pull requests.
 			ctx.Log.Debug("setting VCS status to success with no projects found")
-			if err := c.CommitStatusUpdater.UpdateCombinedCount(baseRepo, pull, models.SuccessCommitStatus, models.PlanCommand, 0, 0); err != nil {
+			if err := c.CommitStatusUpdater.UpdateCombinedCount(ctx.Pull.BaseRepo, ctx.Pull, models.SuccessCommitStatus, cmdModel, 0, 0); err != nil {
 				ctx.Log.Warn("unable to update commit status: %s", err)
 			}
 		}
-		return
+		return err
 	}
```
Comment (nishkrishnan) [inline server/events/command_runner.go:178]: imo this is kind of odd right, because presumably this should be resolved for plan, since policy check always takes place after plan during the auto command.

Maybe this is a more high level PR comment, but after looking through your PR, there is a lot of duplication which is the way it is between plan and apply commands as well. However for policy check, i think of it similar as init, except that it will always follow plan.  So the question I want to know is would it be cleaner if we kept it as a step in the plan command instead? ie. here: https://github.com/lyft/atlantis/blob/ORCA-470-add-policy-check-step-runner/server/events/yaml/valid/global_cfg.go#L83:L85

Benefits I see are the following:
* you don't need to add a separate `parallel*` flag since it will be handled as part of the plan
* you won't be running through unnecessary logic multiple times (building the project config and all that. For projects that are skipping the clone step, it's two downloads of the repo config file.
* A lot less code being written

The only disadvantage I see is that this will make it harder to create a comment command for policy checks specifically, though imo this is ok since it should always follow plan.

Comment (nishkrishnan) [inline server/events/command_runner.go:178]: Presumably, this would already be resolved by the policy checking stage so it's unfortunate we still have to do these kinds of checks given the way commands are built.  Given that policy checking always follows plan, did you consider just adding it as a step in the plan command?  I'm unsure of how complex it is to deal with 2 sets of outputs and 2 commit statuses but the benefits I see are:

* Less duplicated logic to run through (ie. all the project config stuff)
* Less code to write here
* You get parallel flag for free and don't need to create one for policy checking

File: server/events/command_runner.go [hunk 2]
```diff
@@ -134,55 +134,89 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(baseRepo models.Repo, headRepo
...
-	c.updatePull(ctx, AutoplanCommand{}, result)
+	c.updatePull(ctx, pullCommand, result)
 	pullStatus, err := c.updateDB(ctx, ctx.Pull, result.ProjectResults)
 	if err != nil {
 		c.Logger.Err("writing results: %s", err)
 	}
 
-	c.updateCommitStatus(ctx, models.PlanCommand, pullStatus)
+	c.updateCommitStatus(ctx, cmdModel, pullStatus)
```
Comment (nishkrishnan) [inline server/events/command_runner.go:209]: Do you need to update this function?
https://github.com/lyft/atlantis/blob/ORCA-470-add-policy-check-step-runner/server/events/command_runner.go#L380:L381

File: server/events/command_runner.go [hunk 3]
```diff
@@ -601,6 +651,11 @@ func (c *DefaultCommandRunner) parallelPlanEnabled(ctx *CommandContext, projectC
 	return len(projectCmds) > 0 && projectCmds[0].ParallelPlanEnabled
 }
 
+// parallelPolicyCheckEnabled returns true if parallel plan is enabled in this context.
+func (c *DefaultCommandRunner) parallelPolicyCheckEnabled(ctx *CommandContext, projectCmds []models.ProjectCommandContext) bool {
```
Comment (nishkrishnan) [inline server/events/command_runner.go:655]: Should we just only support this, instead of also supporting the flag? Doesn't seem like anyone would care about non-parallel policy checks when there are parallel policy plans or vice versa

**Review by nishkrishnan — COMMENTED** [review]
Nice work, mainly just a high level comment and question to make sure this is the approach we want

File: server/events/models/models.go [hunk 4]
```diff
@@ -442,6 +451,22 @@ type PlanSuccess struct {
...
+// PolicyCheckSuccess is the result of a successful policy check run.
+type PolicyCheckSuccess struct {
+	// PolicyCheckOutput is the output from policy check binary(conftest|opa)
+	PolicyCheckOutput string
+	// LockURL is the full URL to the lock held by this policy check.
+	LockURL string
+	// RePlanCmd is the command that users should run to re-plan this project.
+	RePlanCmd string
+	// ApplyCmd is the command that users should run to apply this plan.
+	ApplyCmd string
```
Comment (nishkrishnan) [inline server/events/models/models.go:463]: If we have policy checks enabled, we should prevent showing the apply command in the plan output. We can do this as a follow up but just calling it out.

File: server/events/models/models.go [hunk 5]
```diff
@@ -442,6 +451,22 @@ type PlanSuccess struct {
...
+	// HasDiverged is true if we're using the checkout merge strategy and the
+	// branch we're merging into has been updated since we cloned and merged
+	// it.
+	HasDiverged bool
```
Comment (nishkrishnan) [inline server/events/models/models.go:467]: Hmm, do we need this? We're only going to be using the plan input right?

File: server/events/models/models.go [hunk 6]
```diff
@@ -505,6 +536,10 @@ func (p ProjectPlanStatus) String() string {
 		return "applied"
 	case DiscardedPlanStatus:
 		return "plan_discarded"
+	case ErroredPolicyCheckStatus:
+		return "policy_check_errored"
+	case PassedPolicyCheckStatus:
```
Comment (nishkrishnan) [inline server/events/models/models.go:541]: Should we differentiate between Passed and Overriden? I would imagine so.

File: server/events/models/models.go [hunk 7]
```diff
@@ -532,6 +569,8 @@ func (c CommandName) String() string {
 		return "plan"
 	case UnlockCommand:
 		return "unlock"
+	case PolicyCheckCommand:
+		return "policy_check"
```
Comment (nishkrishnan) [inline server/events/models/models.go:573]: See previous comment about renaming this to `models.PolicyApprovalCommand`

File: server/events/project_command_builder.go [hunk 8]
```diff
@@ -47,6 +47,10 @@ type ProjectCommandBuilder interface {
 	// comment doesn't specify one project then there may be multiple commands
 	// to be run.
 	BuildApplyCommands(ctx *CommandContext, comment *CommentCommand) ([]models.ProjectCommandContext, error)
+	// BuildPolicyCheckCommands builds project policy_check commands for ctx and
+	// comment. If comment doesn't specify one project then there may be
+	// multiple commands to be run.
+	BuildPolicyCheckCommands(ctx *CommandContext, comment *CommentCommand) ([]models.ProjectCommandContext, error)
```
Comment (nishkrishnan) [inline server/events/project_command_builder.go:53]: I think this can be removed from the interface?

File: server/events/runtime/run_step_runner.go [hunk 9]
```diff
@@ -13,7 +13,7 @@ import (
 
 // RunStepRunner runs custom commands.
 type RunStepRunner struct {
-	TerraformExecutor TerraformExec
+	TerraformExecutor StepCmdExec
```
Comment (nishkrishnan) [inline server/events/runtime/run_step_runner.go:16]: Why did you change this? We should minimize changes and this doesn't seem super essential to do now imo.

File: server/testfixtures/test-repos/automerge/exp-output-auto-policy-check.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+no template matched–this is a bug
```
Comment (nishkrishnan) [inline server/testfixtures/test-repos/automerge/exp-output-auto-policy-check.txt:1]: what's going on here? lol

**Review by nishkrishnan — COMMENTED** [review]
Nice Job! Just a few suggestions and questions and then you should be good to go!

File: server/events/command_runner.go [hunk 11]
```diff
@@ -335,33 +380,42 @@ func (c *DefaultCommandRunner) RunCommentCommand(baseRepo models.Repo, maybeHead
 	if cmd.Name == models.ApplyCommand && c.automergeEnabled(ctx, projectCmds) {
 		c.automerge(ctx, pullStatus)
 	}
+
+	// Runs policy checks step after all plans are successful
+	if cmd.Name == models.PlanCommand &&
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner.go:385]: yeah, I agree. I got somewhat used to if statements using go lol

File: server/events/models/models.go (hunk 6 above)
Comment (msarvar) [reply server/events/models/models.go:541]: We can add that when we start working on that logic. But I agree we should have another status for that.

File: server/testfixtures/test-repos/automerge/exp-output-auto-policy-check.txt (hunk 10 above)
Comment (msarvar) [reply server/testfixtures/test-repos/automerge/exp-output-auto-policy-check.txt:1]: this is the error that you get if you don't have a template specified for the command :D I didn't want to add a template into this PR so I this is a placeholder.

File: server/events/command_runner.go [hunk 12]
```diff
@@ -183,6 +185,42 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(baseRepo models.Repo, headRepo
...
+	if c.parallelPolicyCheckEnabled(ctx, projectCmds) {
+		ctx.Log.Info("Running plans in parallel")
+		result = c.runProjectCmdsParallel(projectCmds, models.PolicyCheckCommand)
+	} else {
+		result = c.runProjectCmds(projectCmds, models.PolicyCheckCommand)
+	}
+
+	c.updatePull(ctx, AutoPolicyCheckCommand{}, result)
+
+	pullStatus, err := c.updateDB(ctx, ctx.Pull, result.ProjectResults)
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner.go:218]: Yes, it stores the statuses of the projects for a given Pull.

File: server/events/command_runner.go [hunk 13]
```diff
@@ -183,6 +185,42 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(baseRepo models.Repo, headRepo
 	}
 
 	c.updateCommitStatus(ctx, models.PlanCommand, pullStatus)
+
+	// Check if there are any planned projects and if there are any errors or if plans are being deleted
+	if len(result.ProjectResults) > 0 &&
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner.go:190]: Why would we trigger policy checks? Wouldn't `ProjectResults` be empty in that case?

File: server/events/command_runner.go [hunk 14]
```diff
@@ -135,22 +135,24 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(baseRepo models.Repo, headRepo
...
 	if len(projectCmds) == 0 {
-		log.Info("determined there was no project to run plan in")
+		ctx.Log.Info("determined there was no project to run plan in")
 		if !c.SilenceVCSStatusNoPlans {
 			// If there were no projects modified, we set a successful commit status
 			// with 0/0 projects planned successfully because some users require
 			// the Atlantis status to be passing for all pull requests.
 			ctx.Log.Debug("setting VCS status to success with no projects found")
-			if err := c.CommitStatusUpdater.UpdateCombinedCount(baseRepo, pull, models.SuccessCommitStatus, models.PlanCommand, 0, 0); err != nil {
+			if err := c.CommitStatusUpdater.UpdateCombinedCount(ctx.Pull.BaseRepo, ctx.Pull, models.SuccessCommitStatus, models.PlanCommand, 0, 0); err != nil {
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner.go:155]: actually, I can revert this back. I had some of this block refactored into helper function and than moved things back and it seems i forgot to revert some code changes as well.

File: server/events/command_runner.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/events/command_runner.go:190]: nah it would still be considered a successful plan, i don't think any planfile introspection occurs in atlantis as only the plan file would say that there are no infrastructure changes.

File: server/events/command_runner.go [hunk 15]
```diff
@@ -335,33 +380,42 @@ func (c *DefaultCommandRunner) RunCommentCommand(baseRepo models.Repo, maybeHead
...
-		numSuccess = len(pullStatus.Projects) - pullStatus.StatusCount(models.ErroredPlanStatus)
-		status = models.SuccessCommitStatus
-		if numSuccess != len(pullStatus.Projects) {
-			status = models.FailedCommitStatus
-		}
-	} else {
+		numSuccess = len(pullStatus.Projects) - numErrored
+	case models.PolicyCheckCommand:
+		numSuccess = pullStatus.StatusCount(models.PassedPolicyCheckStatus)
+		numErrored = pullStatus.StatusCount(models.ErroredPolicyCheckStatus)
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/command_runner.go:406]: Can we? If plan fails we are not gonna run policy checks, so it doesn't mean that policy check passed. Compared to comment above where if apply fails it implies that plan already passed.

[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/events/command_runner.go:406]: hmmm ok, let's move forward with this

**Review by nishkrishnan — APPROVED** [review]
